vm/devices/firmware/firmware_uefi/src/service/diagnostics.rs

```diff
@@ -413,54 +446,65 @@ impl DiagnosticsServices {
 }
 
 impl UefiDevice {
-    /// Process the diagnostics buffer and log the entries to tracing
-    pub(crate) fn process_diagnostics(&mut self) {
+    /// Invokes the diagnostics service to process the diagnostics buffer
+    /// with a handler function that logs the entries to tracing
+    fn process_diagnostics_inner(&self, context: &str) {
+        let mut diagnostics = self.service.diagnostics.lock();
+        if let Err(error) = diagnostics.process_diagnostics(&self.gm, handle_efi_diagnostics_log) {
+            tracing::error!(
+                error = &error as &dyn std::error::Error,
+                context,
+                "failed to process diagnostics buffer"
+            );
+        }
+    }
+
+    /// Process the diagnostics buffer as long as it has not been processed before
+    ///
+    /// NOTE: This is intended to be called in response to the guest issuing the
+    /// PROCESS_EFI_DIAGNOSTICS UefiCommand
+    pub(crate) fn ratelimited_process_diagnostics(&mut self) {
         // Do not proceed if we have already processed before
-        if self.service.diagnostics.did_process {
+        if self.service.diagnostics.lock().did_process {
             tracelimit::warn_ratelimited!("Already processed diagnostics, skipping");
             return;
         }
-        self.service.diagnostics.did_process = true;
+        self.service.diagnostics.lock().did_process = true;
+        self.process_diagnostics_inner("guest");
+    }
+
+    /// Forcefully process the diagnostics buffer regardless of previous state
+    pub(crate) fn force_process_diagnostics(&mut self, context: &str) {
+        self.process_diagnostics_inner(context);
+    }
+}
+
+/// Watchdog callback for diagnostics processing
+pub struct DiagnosticsWatchdogCallback {
+    diagnostics: Arc<Mutex<DiagnosticsServices>>,
```

Comment:
This is really not the way we do things in our codebase. Arc+Mutex is a recipe for deadlocks, circular references, and unclear lifetimes/ownership.

Reply:
A better way that fits with the existing model would probably be to have the callback send a notification to a channel, and then have UEFI's poll_device() wait on that channel and trigger the behavior.